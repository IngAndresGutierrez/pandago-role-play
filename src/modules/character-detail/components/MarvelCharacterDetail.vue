<template>
  <v-container class="py-4">
    <v-row justify="center">
      <v-col cols="12" md="8">

        <v-alert v-if="errorFetch" class="mb-4" prominent type="warning">
          Personaje no encontrado
          <v-btn class="ml-4" color="primary" @click="goHome">
            Regresar
          </v-btn>
        </v-alert>

        <v-card v-if="character">
          <v-container>
            <v-row class="align-center">
              <v-col class="d-flex align-center" cols="2">
                <v-btn icon @click="goBack">
                  <v-icon>mdi-arrow-left</v-icon>
                </v-btn>
              </v-col>

              <v-col class="d-flex justify-center" cols="8">
                <v-card-title>{{ character.name }}</v-card-title>
              </v-col>

              <v-col class="d-flex align-center" cols="2" />
            </v-row>
          </v-container>

          <v-img :alt="character.name" class="mb-4" height="400px" :src="getImageFromThumbnail(character.thumbnail)" />

          <v-card-text v-if="character.description">{{ character.description }}</v-card-text>
          <v-card-text v-else>Descripci√≥n no disponible.</v-card-text>

          <v-card>
            <v-tabs v-model="tab" bg-color="primary">
              <v-tab value="comics">Comics</v-tab>
              <v-tab value="series">Series</v-tab>
              <v-tab value="events">Eventos</v-tab>
              <v-tab value="histories">Historias</v-tab>
            </v-tabs>

            <v-tabs-window v-model="tab">
              <v-tabs-window-item value="comics">
                <MarvelCharacterDetailComics />
              </v-tabs-window-item>

              <v-tabs-window-item value="series">
                <MarvelCharacterDetailSeries />
              </v-tabs-window-item>

              <v-tabs-window-item value="events">
                <MarvelCharacterDetailEvents />
              </v-tabs-window-item>

              <v-tabs-window-item value="histories">
                <MarvelCharacterDetailStories />
              </v-tabs-window-item>
            </v-tabs-window>

          </v-card>
        </v-card>

        <v-skeleton-loader
          v-if="!character && !errorFetch"
          class="mx-auto border"
          max-height="400"
          max-width="400"
          type="image, article"
        />
      </v-col>
    </v-row>
  </v-container>
</template>

<script setup lang="ts">
  import { onMounted } from 'vue'
  import { useRoute, useRouter } from 'vue-router'

  import { Character } from '../models/characters'
  import marvelService from '../services/character'
  import { getImageFromThumbnail } from '../../common/utils/images'
  import MarvelCharacterDetailComics from './MarvelCharacterDetailComics.vue'
  import MarvelCharacterDetailSeries from './MarvelCharacterDetailSeries.vue'
  import MarvelCharacterDetailEvents from './MarvelCharacterDetailEvents.vue'
  import MarvelCharacterDetailStories from './MarvelCharacterDetailStories.vue'

  const route = useRoute()
  const router = useRouter()
  const tab = ref('comics')
  const character = ref<Character | null>(null)
  const errorFetch = ref()

  const fetchCharacterDetail = async () => {
    if (route && route.params && 'id' in route.params) {
      const characterId = Number(route.params.id)

      try {
        const response = await marvelService.getCharacterDetail(characterId)
        if (response) {
          character.value = response
        } else {
          errorFetch.value = 'Personaje no encontrado'
        }
      } catch (err: any) {
        errorFetch.value = err.message || 'Error al obtener el personaje'
      }
    }
  }

  const goBack = () => {
    router.back()
  }

  const goHome = () => {
    router.push('/')
  }

  onMounted(fetchCharacterDetail)

</script>

<style scoped>
.v-card-title {
  font-size: 24px;
  font-weight: bold;

}

.v-img {
  border-radius: 8px;
}
</style>
