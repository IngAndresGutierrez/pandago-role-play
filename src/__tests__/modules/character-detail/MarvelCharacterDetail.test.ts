import 'vuetify/styles'
import { mount } from '@vue/test-utils'
import { createVuetify } from 'vuetify'
import { describe, expect, test } from 'vitest'

import MarvelCharacterDetail from '@/modules/character-detail/components/MarvelCharacterDetail.vue'

const vuetify = createVuetify()

const mockCharacter = {
  id: 1,
  name: 'Test character',
  thumbnail: {
    path: 'https://mypath.com/images',
    extension: 'jpg',
  },
}

describe('MarvelCharacterDetail.vue', () => {
  test('renders correctly', () => {
    const wrapper = mount(MarvelCharacterDetail, {
      global: {
        plugins: [vuetify],
      },
      props: {
        character: mockCharacter,
      },
    })

    expect(wrapper.element).toMatchSnapshot()
  })
})
