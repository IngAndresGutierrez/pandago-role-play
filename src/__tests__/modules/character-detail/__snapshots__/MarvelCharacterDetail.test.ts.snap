// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MarvelCharacterDetail.vue > renders correctly 1`] = `
<div
  character="[object Object]"
  class="v-container v-locale--is-ltr py-4"
  data-v-7908485d=""
>
  <div
    class="v-row justify-center"
    data-v-7908485d=""
  >
    <div
      class="v-col-md-8 v-col-12"
      data-v-7908485d=""
    >
      <div
        aria-label="Loading..."
        aria-live="polite"
        class="v-skeleton-loader v-theme--light mx-auto border"
        data-v-7908485d=""
        role="alert"
        style="max-height: 400px; max-width: 400px;"
      >
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__image"
        >
          
          
        </div>
        
        
        <div
          class="v-skeleton-loader__bone v-skeleton-loader__article"
        >
          
          
          <div
            class="v-skeleton-loader__bone v-skeleton-loader__heading"
          >
            
            
          </div>
          
          
          <div
            class="v-skeleton-loader__bone v-skeleton-loader__paragraph"
          >
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
            <div
              class="v-skeleton-loader__bone v-skeleton-loader__text"
            >
              
              
            </div>
            
            
          </div>
          
          
        </div>
        
        
      </div>
    </div>
  </div>
</div>
`;
